Fix random placement behaviour in new TileMap Editor

snailrhymer opened the pull request:
Fix a couple of bugs in the way tiles were randomly placed that impacted scattering behaviour:

- Correctly calculate probability of not placing a tile when scattering is enabled

- Add tiles chosen as empty when scattering to `drag_modified`
- Only allow changing tiles not in `drag_modified`

Together, the last two fix a bug where scattering could be set to e.g. 0.9 but a tile would still almost always be placed, as the check `rand > 0.9` would be run on every mouse movement until it was successful.

***

Arguably, the changes make the name `drag_modified` inaccurate. It could be renamed, or a second HashSet could be added to separately track "don't write to during this drag"-type behaviour.

snailrhymer committed:
Fix random placement behaviour in new TileMap Editor

Fix a bug where scattering could be set to e.g. 0.9 but a tile would still always be placed, as the check rand > 0.9 would be run on every mouse movement until it was successful.

groud commented:
I am not 100% convinced by this PR. I get why you would prefer that a stroke does not paint on areas you already dragged upon, 
but I feel like users would complain about the fact it would not paint anything on tile they would insist on (like, moving the mouse a lot on the same tile). Though I agree it would respect the scattering parameter more, so maybe it's ok to ask users to click again if they need to. But I am not sure it is better.

But, in any case, on that matter, if this behavior is decided as better, the code is fine.

> Together, the last two fix a bug where scattering could be set to e.g. 0.9 but a tile would still almost always be placed, as the check rand > 0.9 would be run on every mouse movement until it was successful.

This is expected. The scattering parameter is an increasing value. The higher it is, the more spread apart the tiles should be. Like, a value of 1000 would paint one tile every 1000 items. I think it is a lot easier to use than a percentage value, as it is a lot easier to type and manipulate "1000" than "0.001".

Changing this should be part of a proposal.

snailrhymer commented:
**A**

> The scattering parameter is an increasing value. The higher it is, the more spread apart the tiles should be. Like, a value of 1000 would paint one tile every 1000 items. I think it is a lot easier to use than a percentage value, as it is a lot easier to type and manipulate "1000" than "0.001".
> 
> Changing this should be part of a proposal.

I'm happy to discuss the _scale_ of the scattering parameter as a proposal, but the purpose of the changes to `_pick_random_tile` was more to address the fact that the proportion of empty tiles placed was dependent not only on the scatter parameter, but also the sum of tile weights. If I'm drawing with some set of `used_cells` with total weight `k`, then `empty_probability = k * scattering` and `rand` is uniform in the range `[0, k + k * scattering]`. So the probability of no tile being placed is `k/(k * (1 + s))`. (Sorry if I'm pointing out the obvious, I just want to make sure we're on the same page.)

So e.g. if I have only one tile with weight 1.0 in my `used_cells`, I need to choose `scattering` as `2.0` to get 33% coverage. Then if I add another tile, I need to change scattering to `4.0` to get the same 33% coverage. I get the logic of this, acting as if there's a virtual empty tile with weight `scattering` in the set of tiles to choose from, but it feels like a lot of mental arithmetic for the user (especially once you have an atlas with many non-integer weights, or are changing the `used_cells` with any regularity). I think that having the effects of scattering be independent of `used_cells` would be more convenient.

***

**B**

> I get why you would prefer that a stroke does not paint on areas you already dragged upon, but I feel like users would complain about the fact it would not paint anything on tile they would insist on (like, moving the mouse a lot on the same tile). Though I agree it would respect the scattering parameter more, so maybe it's ok to ask users to click again if they need to. But I am not sure it is better.

If I set up a scatter parameter that gives me e.g. 50% fill (whatever value I need to achieve that) when I use the rectangle or line tools, when I use the pencil of bucket fill tools, I will invariably get very near to 100% coverage. This coverage depends on how quickly I move the mouse as I drag. This means that if I want to draw a freehand line with a given coverage, I need to trial and error to the correct scatter parameter for a given mouse movement speed and then draw my line with constant arc length.

I'm not entirely happy with the solution I landed on, as I agree that it'd be nice to be able to revisit areas to fill them in more within a single drag, but I think it would lead to better usability of using scatter with the pencil tool.

An alternative I considered was to only prevent "rerolling" the tile while the mouse moved within a single grid space. This would mean that fill % would be consistent, but users could also flail their mouse around over an area until they were happy with the fill. The downside of this approach is that it doesn't really have a meaningful equivalent when it comes to clicking and dragging with bucket fill, but that that's probably a very rare use case.

***

To summarise my wall of text, I think there are four distinct issues:

1. The scale of the `scattering` parameter.
2. Whether the `scattering` parameter should represent a _probability_ of a tile being empty or the _weight_ of an empty tile.
3. Click and drag behaviour for the pencil tool and how it relates to scattering.
4. Click and drag behaviour for the bucket fill tool and how it relates to scattering.

And I'd propose resolving them:

1. Discuss the scale of the parameter separately as a proposal.
2. If you disagree with my points in **A**, move that concept to its own proposal as well.
3. Change the pencil + scatter behaviour so that mouse movements within a tile don't cause the tile to get rerolled, but otherwise revisiting previous tiles does. (This is a change I'm confident I can implement, so hopefully you can forgive my lack of implementation details.)
4. Keep the changes for clicking and dragging with the bucket fill tool.

groud commented:
> So e.g. if I have only one tile with weight 1.0 in my used_cells, I need to choose scattering as 2.0 to get 33% coverage. Then if I add another tile, I need to change scattering to 4.0 to get the same 33% coverage. I get the logic of this, acting as if there's a virtual empty tile with weight scattering in the set of tiles to choose from, but it feels like a lot of mental arithmetic for the user (especially once you have an atlas with many non-integer weights, or are changing the used_cells with any regularity). I think that having the effects of scattering be independent of used_cells would be more convenient.

I understand your point, but the problem would be that, if we want to make it "arithmetically" better, it means that we would have to set the minimum value to one, and the probability to place a tile would become 1 out of `scattering value` (1 would need to be the minimum as you cannot place more often than 1 out of 1 tile).

The issue I would have with that, is that I think it would be more confusing than the current state, as 0 means "no scattering" and higher values means "more scattering". It may be a little bit more complex to have a precise probability values, but I am pretty sure no one needs precise value here. That's why it requires a proposal if it needs to be discussed.

An alternative may be to display the percentage somewhere, so that editing the scattering value also tells you the percentage probability to place a tile. Like in a tooltip or something like that.

> If I set up a scatter parameter that gives me e.g. 50% fill (whatever value I need to achieve that) when I use the rectangle or line tools, when I use the pencil of bucket fill tools, I will invariably get very near to 100% coverage.

Ah I see, this is an annoying issue that likely outweighs the problem of having to click again on an empty tile if it did not fill in.
I guess the change is worth merging then. We will see if anyone complains about this new behavior if it causes any problem.

snailrhymer commented:
Sorry, I've realised I was wrong about the scattering behaviour depending on the sum of weights of tiles and so have probably made what I intended a lot less clear. I still think it's worth changing the relationship between `scattering` and placement probability to be linear: godotengine/godot-proposals#4591.

> if we want to make it "arithmetically" better, it means that we would have to set the minimum value to one, and the probability to place a tile would become 1 out of `scattering value`

I don't think I agree that this is a consequence my proposed change; hopefully I've made what I'm suggesting clearer in the proposal.

> An alternative may be to display the percentage somewhere, so that editing the scattering value also tells you the percentage probability to place a tile. Like in a tooltip or something like that.

If the current system is kepy, I think that some display of the percentage would be a useful addition.

***

> I guess the change is worth merging then.

Do you think it's worth altering it to allow revisiting tiles within one drag, i.e. only preventing the "rerolls" when the mouse is moving inside of one grid square?

snailrhymer commented:
Removed changes to how the `scattering` parameter determines the probability of no tile being placed. Added a change to the tooltip hopefully making it clearer that `scattering` is not placement probability.

groud reviewed (approved):
Sorry for the very late review. The PR needs a rebase but it looks good to me.

akien-mga commented:
I pushed a rebase, but it would be good if someone could test the PR and confirm that it works as expected.

akien-mga commented:
Thanks!
